### ISSUE TITLE
Use NeuroML2 ComponentTypes in isopotential.xml

### ISSUE THREAD

**Issue by jrieke** [cengique/drosophila-aCC-L3-motoneuron-model#4]
@pgleeson @borismarin I added a first version of the simple isopotential model in NeuroML2 (see [isopotential.xml](https://github.com/cengique/drosophila-aCC-L3-motoneuron-model/blob/master/NeuroML2/isopotential.xml)). Results look good:

![image](https://cloud.githubusercontent.com/assets/5103165/9618309/d961c68e-5105-11e5-835b-e29f07ab591a.png)

Right now it's mostly just basic LEMS - I haven't looked at the ComponentTypes of NeuroML2 in detail yet. I actually wanted to use them for a next version; does that make sense for this model (I have seen that the simple version of the Pinsky Rinzel model also uses pretty basic LEMS rather than NeuroML2, so I am a bit unsure)? 

Also, I had to do a little hack (see lines 14-16) in the current version to get units into the value="..." fields. If I use units directly in those fields (e. g. if do something like ```<DerivedVariable ... value="...V-10mV...">```), jnml throws a NumberFormatException. Is this intended or a bug?

**Comment by pgleeson** [comment]
Looks good so far @jrieke!

Have added an issue here (#5) to help automatically compare this model to XPP version.

For your questions, yes, it would be good to break this out into a fully NeuroML 2 based version. This [familiar friend](https://github.com/openworm/muscle_model/tree/master/NeuroML2) should provide a good basis for that. Keep this pure LEMS version for reference.

Yes, as you saw you can't write expressions like v -10mV in expressions, it's a feature, not bug. What you've done with 17.55*in_mV is fine, but better would be to define constants for all of these e.g. 

    <Constant name="minfKf_vhalf" dimension="voltage" value="17.55 mV"/>

and use that in the expression. Alternatively, when you define individual files for the conductances, there are some in built ways to express these in channel files, e.g. `<steadyState type="HHSigmoidVariable" ...>`: 
https://github.com/OpenSourceBrain/L5bPyrCellHayEtAl2011/blob/master/neuroConstruct/generatedNeuroML2/K_Pst.channel.nml

**Comment by cengique** [comment]
@jrieke The results in the screenshot looks fine to me! However, if you want to test the limits of the model, you may want to see if it exhibits the *delay-to-first-spike* property when you lower in the stimulus current close to the firing threshold.

**Comment by borismarin** [comment]
@jrieke shall we open an issue to discuss the dynamics of this model
(steady state / hkf2)? Or would you prefer to work in a separate repo?

**Comment by borismarin** [comment]
@jrieke @cengique see #6

**Comment by jrieke** [comment]
@pgleeson I added a first (unfinished) version of the [pure NeuroML2 file](https://github.com/cengique/drosophila-aCC-L3-motoneuron-model/blob/master/NeuroML2/isopotential.nml) and a short [LEMS file to run it](https://github.com/cengique/drosophila-aCC-L3-motoneuron-model/blob/master/NeuroML2/run_isopotentialnml.xml). I used ```pointCellCondBased``` rather than ```cell``` as in the OpenWorm muscle model you linked because there is no morphology associated with the original model. 

Problems:

- [ ] Time constants for the gating variables: As far as I can see, there is only ```fixedTimeCourse``` available currently, but the time constants in the original model are mostly voltage dependent (sigmoid functions with an offset).
- [ ] Is there any possibility to define initial values for the gating variables?
- [ ] The fast K channel has two different values/equations for the steady state variable of the h gate (```hinfk``` and ```hinfk2```), which get added like ```(fh*hKf + (1-fh)*hKf2)``` for the final value of the gate. Any idea how to implement this?

@borismarin Let's keep the hkf2 issue here, as it might be a bug in the model. For the detailed dynamics, I guess a separate repo might be better (will create it once I get to that). 

@cengique Will definitely try this out - I will investigate the dynamics of the model in quite some detail during the coming weeks because I am writing my bachelor thesis in physics about this.

**Comment by cengique** [comment]
@jrieke regarding the hkf2 issue, it's not a bug at this point, but a feature. However, Ikf and, especially hkf2, don't do much in the model, so you could easily ignore it in the first NeuroML2 file. So you could completely remove hkf2. But I did see @borismarin found some interesting dynamics related to it.

**Comment by pgleeson** [comment]
#### Time constants for the gating variables:

See this: https://github.com/OpenSourceBrain/Thalamocortical/blob/master/neuroConstruct/generatedNeuroML2/ka.channel.nml for an example of an nml2 file using a generic expression for the time course

#### initial values for the gating variables:

These are by default set to the inf/steady state value at t=0 for the starting membrane potential. see defn here (q = inf): https://www.neuroml.org/NeuroML2CoreTypes/Channels.html#gateHHtauInf, https://github.com/NeuroML/NeuroML2/blob/master/NeuroML2CoreTypes/Channels.xml#L253

You could in theory make a new component type for a gate that has an arbitrary initial value for q if you really want to reproduce the xpp

#### The fast K channel has two different values/equations for the steady state variable

You can also make a generic expression for the inf, e.g. https://github.com/OpenSourceBrain/L5bPyrCellHayEtAl2011/blob/master/neuroConstruct/generatedNeuroML2/SK_E2.channel.nml


Also, I'd recommend breaking the model up into *.cell.nml and *.channel.nml files & including these. Easier to plot the activation variable functions then with pynml-channelanalysis (see http://www.opensourcebrain.org/docs#Converting_To_NeuroML2)

**Comment by jrieke** [comment]
@pgleeson I implemented the custom time constants, the inf for the fast K channel, and split everything up into several files.

I also tried making a new ComponentType to set initial values for the gates, but in my eyes it adds more overhead to the model than what it's useful for (the spike times differ <0.03 ms from the original model with the default ComponentTypes/gates).

So, basically there are 3 options:

1. We leave it like it is (with the default gates where the gating variables get set to inf on start), I add a tolerance of 0.03ms in the omv test and include a comment in the .nml file that the values differ slightly
2. I add the custom ComponentTypes to make it 100% according to the original model (but as mentioned it adds a lot of complexity)
3. We change the init values in the original isopotential.ode file to the inf values of the gates - @cengique would you be OK with that or does anything speak against this step?

I would prefer option 3 if it's possible, what do you guys think?

**Comment by cengique** [comment]
@jrieke I'm OK with #3 above. They are usually set from steady state conditions that I achieve by running XPP for a long time using desired input currents.

**Comment by pgleeson** [comment]
I'd suggest going for option3 too, but I'd leave the original  isopotential.ode pretty much unchanged (as it was the one used in the publication), and make a new isopotential2.ode (or something), with your modifications/optimisations/lots of comments etc. that you can change at will (keeping the overall behaviour as close as possible to the original).

See also pull request https://github.com/cengique/drosophila-aCC-L3-motoneuron-model/pull/9

**Comment by cengique** [comment]
I kind of agree with leaving isopotential.ode in pristine state. However the model will eventually evolve. So how about we use the tag or branch features to freeze the paper model? A tag/release may be more appropriate.

**Comment by jrieke** [comment]
First of all, I made a release with all original files as @cengique suggested ([v1.0](https://github.com/cengique/drosophila-aCC-L3-motoneuron-model/releases/tag/v1.0)). 

Secondly, I found a slight mistake in how I handled hKf2 in the NeuroML2 version, fixed this now. Also, I made all isopotential models (XPP, LEMS, NeuroML2) use the init values for the gates calculated from the respective inf values (as outlined above). However, the spike times of the NeuroML2 version are still slightly off from the XPP and LEMS version (see the [test results](https://travis-ci.org/cengique/drosophila-aCC-L3-motoneuron-model#L317)). The required tolerance to pass the test has reduced to 0.00120 from 0.00241, but it's still weirdly off - and I cannot figure out why. I already checked:

- The init values get correctly set in all cases (compared the output in the .dat file to the data browser in XPP)
- The values for potential and gating variables are perfectly equal for the first 10 ms; as soon as the current step begins, they slightly start to diverge from each other
- The curves of the potential and all gating variables look similar in the graphs
- All parameter values are correct (triple checked this)

@pgleeson I doubt it as the LEMS version works fine, but could this have anything to do with  numerical issues or how jnml handles calculations in the NeuroML2 ComponentTypes?

Finally, sorry for the long delay, I was travelling and then spent the last three days trying to figure out why in god's name the spike times are different :)
